### PR TITLE
update prep_sample

### PR DESCRIPTION
**Pull request by edsu7**
- added supported for single end and paired end sequencing experiment detection
- added check for local files and score IDs (analysis and study)
- moved JSON attachment from python script to outer groovy script

### CODE REVIEW

File: modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py [hunk 1]
```diff
@@ -262,9 +262,17 @@ def main():
...
+    if song_analysis.get('workflow'):
+      genome_build = song_analysis['workflow']["genome_build"]
+    else:
+      genome_build = None
+
     analysis_type = song_analysis['analysisType']['name']
     output_sample_sheet = f'{args.outdir}/{sample_id}_{analysis_type}_sample_sheet.csv'
-    
+    experiment=song_analysis['experiment']['experimental_strategy']
+    output_sample_sheet = f'{args.outdir}/{sample_id}_{analysis_type}_sample_sheet.csv'
```
Comment (lindaxiang) [inline modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py:274]: There is a duplicated definition for `output_sample_sheet`?

File: modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py [hunk 2]
```diff
@@ -337,8 +345,8 @@ def main():
           sys.exit("Error: not supported input file format")
       with open(output_sample_sheet, 'w', newline='') as f:
         csvwriter = csv.writer(f, delimiter=',')
-        csvwriter.writerow(['analysis_type','study_id','patient','sex','status','sample','cram','crai', 'analysis_json'])
-        csvwriter.writerow([analysis_type, study_id, donor_id, sex, status, sample_id, cram, crai, metadata_json])
+        csvwriter.writerow(['analysis_type','study_id','patient','sex','status','sample','cram','crai',"genome_build",'experiment'])
+        csvwriter.writerow([analysis_type, study_id, donor_id, sex, status, sample_id, cram, crai, experiment,genome_build, metadata_json])
```
Comment (lindaxiang) [inline modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py:349]: - `experiment` and `genome_build` should switch position?
- `metadata_json` is missing field definition

File: modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py [hunk 3]
```diff
@@ -355,13 +363,13 @@ def main():
           sys.exit("Error: not supported input file format")
       with open(output_sample_sheet, 'w', newline='') as f:
         csvwriter = csv.writer(f, delimiter=',')
-        csvwriter.writerow(['analysis_type','study_id','patient','sex','sample','variantcaller','vcf','tbi', 'analysis_json'])
-        csvwriter.writerow([analysis_type, study_id, donor_id, sex, sample_id, variantcaller, vcf, tbi, metadata_json]) 
+        csvwriter.writerow(['analysis_type','study_id','patient','sex','sample','variantcaller','vcf','tbi',"genome_build",'experiment'])
+        csvwriter.writerow([analysis_type, study_id, donor_id, sex, sample_id, variantcaller, vcf, tbi ,experiment,genome_build, metadata_json])
```
Comment (lindaxiang) [inline modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py:367]: same here:
- `experiment` and `genome_build` should switch position?
- `metadata_json` is missing field definition

File: modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py [hunk 4]
```diff
@@ -355,13 +363,13 @@ def main():
...
-        csvwriter.writerow(['analysis_type','study_id','patient','sex','sample','variantcaller','vcf','tbi', 'analysis_json'])
-        csvwriter.writerow([analysis_type, study_id, donor_id, sex, sample_id, variantcaller, vcf, tbi, metadata_json]) 
+        csvwriter.writerow(['analysis_type','study_id','patient','sex','sample','variantcaller','vcf','tbi',"genome_build",'experiment'])
+        csvwriter.writerow([analysis_type, study_id, donor_id, sex, sample_id, variantcaller, vcf, tbi ,experiment,genome_build, metadata_json])  
 
     elif analysis_type == 'qc_metrics':
       with open(output_sample_sheet, 'w', newline='') as f:
         csvwriter = csv.writer(f, delimiter=',')
-        csvwriter.writerow(['analysis_type','study_id','patient','sex','status','sample','qc_tools','qc_file', 'analysis_json'])
+        csvwriter.writerow(['analysis_type','study_id','patient','sex','status','sample','qc_tools','qc_file','experiment','analysis_json'])
```
Comment (lindaxiang) [inline modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py:372]: we can also add field `genome_build` here for qc_files

File: modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py [hunk 5]
```diff
@@ -370,8 +378,7 @@ def main():
             os.symlink(os.path.abspath(fp), qc_file)
             qc_tools = ','.join(fq['info']['analysis_tools'])
 
-          csvwriter.writerow([analysis_type, study_id, donor_id, sex, status, sample_id, qc_tools, qc_file, metadata_json]) 
-
+          csvwriter.writerow([analysis_type, study_id, donor_id, sex, status, sample_id, qc_tools, qc_file,experiment, metadata_json])
```
Comment (lindaxiang) [inline modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py:381]: we can also add field `genome_build` here for qc_files

**Review by lindaxiang — CHANGES_REQUESTED** [review]

File: modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py [hunk 6]
```diff
@@ -262,9 +262,16 @@ def main():
...
     status = '0' if tumour_normal_designation == 'Normal' else '1'
+    
+    if song_analysis.get('workflow'):
+      genome_build = song_analysis['workflow']["genome_build"]
+    else:
+      genome_build = None
+
     analysis_type = song_analysis['analysisType']['name']
     output_sample_sheet = f'{args.outdir}/{sample_id}_{analysis_type}_sample_sheet.csv'
-
```
Comment (edsu7) [inline modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py:267]: Addressing : 
https://github.com/icgc-argo-workflows/argo-modules/pull/22#discussion_r1409446996

File: modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py [hunk 7]
```diff
@@ -337,8 +344,8 @@ def main():
           sys.exit("Error: not supported input file format")
       with open(output_sample_sheet, 'w', newline='') as f:
         csvwriter = csv.writer(f, delimiter=',')
-        csvwriter.writerow(['analysis_type','study_id','patient','sex','status','sample','cram','crai', 'analysis_json'])
-        csvwriter.writerow([analysis_type, study_id, donor_id, sex, status, sample_id, cram, crai, metadata_json])
+        csvwriter.writerow(['analysis_type','study_id','patient','sex','status','sample','cram','crai',"genome_build",'experiment', 'analysis_json'])
+        csvwriter.writerow([analysis_type, study_id, donor_id, sex, status, sample_id, cram, crai, genome_build,experiment, metadata_json])
```
Comment (edsu7) [inline modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py:348]: Addressing:
https://github.com/icgc-argo-workflows/argo-modules/pull/22#discussion_r1409448688

File: modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py [hunk 8]
```diff
@@ -355,13 +362,13 @@ def main():
           sys.exit("Error: not supported input file format")
       with open(output_sample_sheet, 'w', newline='') as f:
         csvwriter = csv.writer(f, delimiter=',')
-        csvwriter.writerow(['analysis_type','study_id','patient','sex','sample','variantcaller','vcf','tbi', 'analysis_json'])
-        csvwriter.writerow([analysis_type, study_id, donor_id, sex, sample_id, variantcaller, vcf, tbi, metadata_json]) 
+        csvwriter.writerow(['analysis_type','study_id','patient','sex','sample','variantcaller','vcf','tbi',"genome_build",'experiment', 'analysis_json'])
+        csvwriter.writerow([analysis_type, study_id, donor_id, sex, sample_id, variantcaller, vcf, tbi ,genome_build,experiment, metadata_json])
```
Comment (edsu7) [inline modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py:366]: Addressing:
https://github.com/icgc-argo-workflows/argo-modules/pull/22#discussion_r1409451082

File: modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py [hunk 9]
```diff
@@ -355,13 +362,13 @@ def main():
...
-        csvwriter.writerow(['analysis_type','study_id','patient','sex','sample','variantcaller','vcf','tbi', 'analysis_json'])
-        csvwriter.writerow([analysis_type, study_id, donor_id, sex, sample_id, variantcaller, vcf, tbi, metadata_json]) 
+        csvwriter.writerow(['analysis_type','study_id','patient','sex','sample','variantcaller','vcf','tbi',"genome_build",'experiment', 'analysis_json'])
+        csvwriter.writerow([analysis_type, study_id, donor_id, sex, sample_id, variantcaller, vcf, tbi ,genome_build,experiment, metadata_json])  
 
     elif analysis_type == 'qc_metrics':
       with open(output_sample_sheet, 'w', newline='') as f:
         csvwriter = csv.writer(f, delimiter=',')
-        csvwriter.writerow(['analysis_type','study_id','patient','sex','status','sample','qc_tools','qc_file', 'analysis_json'])
+        csvwriter.writerow(['analysis_type','study_id','patient','sex','status','sample','qc_tools','qc_file',"genome_build", 'experiment','analysis_json'])
```
Comment (edsu7) [inline modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py:371]: Addressing:
https://github.com/icgc-argo-workflows/argo-modules/pull/22#discussion_r1409452406

File: modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py [hunk 10]
```diff
@@ -370,8 +377,7 @@ def main():
             os.symlink(os.path.abspath(fp), qc_file)
             qc_tools = ','.join(fq['info']['analysis_tools'])
 
-          csvwriter.writerow([analysis_type, study_id, donor_id, sex, status, sample_id, qc_tools, qc_file, metadata_json]) 
-
+          csvwriter.writerow([analysis_type, study_id, donor_id, sex, status, sample_id, qc_tools, qc_file, genome_build, experiment, metadata_json])
```
Comment (edsu7) [inline modules/icgc-argo-workflows/prep/sample/resources/usr/bin/main.py:380]: Addressing:
https://github.com/icgc-argo-workflows/argo-modules/pull/22#discussion_r1409453776

**Review by lindaxiang — APPROVED** [review]
LGTM

**Pull request merged**